### ISSUE TITLE
Unable to listen on a linux bridge

### ISSUE THREAD

**Issue by gabrik** [haesbaert/rawlink#12]
I have a strange issue when using this library to listen on a Linux Bridge interface

The code I wrote is pretty simple
```
let create_socket iface =
  Lwt.return @@ Lwt_rawlink.open_link iface

let sock_task sock =
  let%lwt _ = Lwt_io.printf "Reading from socket...\n" in
  Lwt_rawlink.read_packet sock
  >>= fun buf ->
  let%lwt _ = Lwt_io.printf ">> read %d bytes\n" (Cstruct.len buf) in
  let%lwt hdr = Eth.from_pkt buf in
  Lwt.return @@ SockTask hdr

let rec loop st sock =
  let%lwt _ = Lwt_io.printf "In loop...\n" in
  Lwt.choose [st sock]
  >>= fun res ->
  match res with
  | SockTask s ->
    let%lwt _ = Lwt_io.printf ">> Addresses DST:%s SRC: %s\n" (Macaddr.to_string s.dst) (Macaddr.to_string s.src) in
    loop st sock
  | _ -> failwith "Error, unkown task result type"

let main _ =
  let%lwt _ = Lwt_io.printf "Started...\n" in
  let face = "br0" in
  let%lwt sock  = create_socket face in
  let%lwt _ = Lwt_io.printf "Socket open...\n" in
  loop sock_task sock


let () = Lwt_main.run @@ main ()
```

Inside the module Eth I simply extract the address from the Ethernet header,
tha strange thing is that if I use a tcpdump on the same interface the code above works, otherwise it simply not work.

Any suggestion?

**Comment by haesbaert** [comment]
You mean if you leave tcpdump running in the background you see packets ?

If yes,
tcpdump sets the interface to promisc, which rawlink does not by default.

It's possible we want to add a function to enable promiscuous mode, if you can confirm that running tcpdump makes you see packets and as soon as you close tcpdump they stop, would be nice.

**Comment by gabrik** [comment]
@haesbaert yes, sorry for not mentioning it, but you got exactly the point, it works if tcpdump is running and when I close tcpdump I cannot see packets anymore.
I think a function to enable promiscuous mode would be useful

Should be enough open the socket using the PF_PACKET instead of AF_PACKET right?

**Comment by haesbaert** [comment]
> Should be enough open the socket using the PF_PACKET instead of AF_PACKET right?

Not really, PF_PACKET and AF_PACKET is the same thing, it's just a historical thing.
We have to do a SIOCGIFFLAGS and SIOCSIFFFLAGS (get and set flags) for IFFPROMISC.

I think an optional argument for open is good enough, I'll see if I can cook something this week, it's been a while and I have to update my development machine.

As a temporary work-around for your case, you can turn the interface into promiscuous mode by ifconfig on linux, I don't know the magic but you can figure it out. If you can do this would be great so we can be sure setting the interface to promiscuous would help your case.

**Comment by gabrik** [comment]
Just tested, setting the interface in promiscuous allow to see packets.
The magic is pretty easy, just 
```
# ip link set br0 promisc on
```

Yes an option argumento to the open would be perfect, in the meanwhile I will work with promiscuous mode on

**Comment by haesbaert** [comment]
> Just tested, setting the interface in promiscuous allow to see packets.
The magic is pretty easy, just

Awesome, I'll add an option to open_link, something like ~promisc:true.
Or, if you wanna write it, I'll happily review it too ;=)

**Comment by gabrik** [comment]
I'm not so good in C so I prefer to not write it, otherwise it will take ages :)

**Comment by gabrik** [comment]
@haesbaert actually it was easier than I thought, but this works only on Linux, I have no way to test this on other UNIXes

**Comment by haesbaert** [comment]
> @haesbaert actually it was easier than I thought, but this works only on Linux, I have no way to test this on other UNIXes

Cool, why did you re-indent the whole file and didn't follow the indentation style ?
I'll pick this up and massage a bit, we also have to restore the flag when the descriptor is closed, the BPF parts I can write too.

**Comment by gabrik** [comment]
@haesbaert I think the indent was changed by VSCode when saving the file, really sorry for that, I did not noticed

**Comment by haesbaert** [comment]
Fixed in b7284a9
